### PR TITLE
Created a general function for retrying commands

### PR DESCRIPTION
**Pull request by JennyChenyj**
[skip ci]

Signed-off-by: Jenny Chen <Jenny.Chen@ibm.com>

### CODE REVIEW

**Comment by JennyChenyj** [comment]
Hi @AdamBrousseau. Thanks!

**Comment by JennyChenyj** [comment]
Hi @AdamBrousseau! The order has been changed. Thanks!

File: buildenv/jenkins/common/pipeline-functions.groovy [hunk 1]
```diff
@@ -606,6 +606,18 @@ def generate_test_jobs(TARGET_NAMES, SPEC, ARTIFACTORY_SERVER, ARTIFACTORY_REPO)
...
 
+def retry_and_delay(command, numRetries = 3, waitTime = 60, units = 'SECONDS') {
+    def ret = false
+    retry(numRetries) {
+    if (ret) {
+        sleep time: waitTime,unit: units
+    } else {
+        ret = true
+    }
+    command()
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/pipeline-functions.groovy]: Indenting needs to be fixed

File: buildenv/jenkins/common/pipeline-functions.groovy [hunk 2]
```diff
@@ -606,6 +606,18 @@ def generate_test_jobs(TARGET_NAMES, SPEC, ARTIFACTORY_SERVER, ARTIFACTORY_REPO)
     }
 }
 
+def retry_and_delay(command, numRetries = 3, waitTime = 60, units = 'SECONDS') {
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/pipeline-functions.groovy]: We should add a comment about how to use this function.

File: buildenv/jenkins/common/pipeline-functions.groovy [hunk 3]
```diff
@@ -606,6 +606,18 @@ def generate_test_jobs(TARGET_NAMES, SPEC, ARTIFACTORY_SERVER, ARTIFACTORY_REPO)
...
+    def ret = false
+    retry(numRetries) {
+    if (ret) {
+        sleep time: waitTime,unit: units
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/pipeline-functions.groovy]: Whitepsace after comma.

Comment (JennyChenyj) [reply buildenv/jenkins/common/pipeline-functions.groovy]: Thanks for find those mistakes! I have fixed them. Thanks for checking it again~

**Comment by pshipton** [comment]
They are merged now.

**Comment by JennyChenyj** [comment]
> They are merged now.

Thanks!

**Comment by JennyChenyj** [comment]
Hey @AdamBrousseau! I replaced both of them. Thanks for checking! :)

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by JennyChenyj** [comment]
Hey @AdamBrousseau! Sorry, I made a mistake about using the function name. I fixed it. Thanks!

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

File: buildenv/jenkins/common/variables-functions.groovy [hunk 4]
```diff
@@ -1044,15 +1044,7 @@ def create_job(JOB_NAME, SDK_VERSION, SPEC, downstreamJobType, id){
...
-    ret = false    
-    retry(3) {
-       if (ret) {
-          sleep time: 120,unit: 'SECONDS'
-       } else {
-          ret = true
-       }
-       jobDsl targets: templatePath, ignoreExisting: false, additionalParameters: params
-    }
+    retry_and_delay({jobDsl targets: templatePath, ignoreExisting: false, additionalParameters: params}, 3, 120)
```
Comment (AdamBrousseau) [inline buildenv/jenkins/common/variables-functions.groovy]: You'll have to check if `pipeline-functions.groovy` is loaded at this point in the build and what variable it's loaded into. Same comment for build.groovy. Look for a `load` step
ex `buildFile = load 'buildenv/jenkins/common/pipeline-functions.groovy'`

Comment (JennyChenyj) [reply buildenv/jenkins/common/variables-functions.groovy]: Hey Adam! I think ```pipeline-functions.groovy``` is not loaded. I checked the console output and there is no line with load. Thanks!

Comment (AdamBrousseau) [reply buildenv/jenkins/common/variables-functions.groovy]: Right, so we'll have to make sure we load that library in those files so we can utilize this new function.

**Comment by JennyChenyj** [comment]
Hey @AdamBrousseau! Thanks for checking it!

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by AdamBrousseau** [comment]
Jenkins compile plinux jdk11

**Comment by pshipton** [comment]
The commits can be squashed, I don't think we need to have all three merged. Also, a different formatting for `pipelineFunctions.retry_and_delay({server.upload spec: uploadSpec, buildInfo: buildInfo; server.publishBuildInfo buildInfo}, 3, 300)` should be easier to read. Something like
```
pipelineFunctions.retry_and_delay({
    server.upload spec: uploadSpec, buildInfo: buildInfo;
    server.publishBuildInfo buildInfo},
    3, 300)
```

**Comment by JennyChenyj** [comment]
> The commits can be squashed, I don't think we need to have all three merged. Also, a different formatting for `pipelineFunctions.retry_and_delay({server.upload spec: uploadSpec, buildInfo: buildInfo; server.publishBuildInfo buildInfo}, 3, 300)` should be easier to read. Something like
> 
> ```
> pipelineFunctions.retry_and_delay({
>     server.upload spec: uploadSpec, buildInfo: buildInfo;
>     server.publishBuildInfo buildInfo},
>     3, 300)
> ```

Thanks for the advice! I will make the change now.

**Comment by JennyChenyj** [comment]
Hi @pshipton! I have made the change. Thanks!

**Pull request merged**